chatkit-core/src/main/kotlin/com/pusher/chatkit/ChatkitTokenProvider.kt

```diff
@@ -41,56 +42,45 @@ class ChatkitTokenProvider
         tokenCache.clearCache()
     }
 
-    private fun fetchTokenFromEndpoint(tokenParams: Any?): Future<Result<String, Error>> {
-        val urlBuilder = HttpUrl.parse(endpoint)!!
-            .newBuilder()
-        urlBuilder.addQueryParameter("user_id", userId)
-
-        val requestBodyBuilder = FormBody.Builder()
-            .add("grant_type", "client_credentials")
-
-        //TODO: Maybe add to query params instead???
-        //Add any extras to the token provider's request.
-        if (tokenParams is ChatkitTokenParams) {
-            tokenParams.extras.keys.forEach { key ->
-                requestBodyBuilder.add(key, tokenParams.extras.getValue(key))
+    private fun fetchTokenFromEndpoint(tokenParams: Any?): Future<Result<String, Error>> =
+        httpUrl.toFuture().flatMapFutureResult { url ->
+            Futures.schedule {
+                val request = Request.Builder().apply {
+                    url(url)
+                    post(requestBody(tokenParams))
+                }.build()
+                val response = request.let { client.newCall(it).execute() }
+                when {
+                    response.isSuccessful && response.code() in 200..299 -> parseTokenResponse(response)
+                    else -> response.asError().asFailure()
+                }
             }
         }
-        //TODO: Maybe add to query params instead???
-        authData.keys.forEach { key ->
-            requestBodyBuilder.add(key, authData.getValue(key))
-        }
 
-        val requestBody = requestBodyBuilder.build()
+    private val httpUrl: Result<HttpUrl, Error>
+        get() = HttpUrl.parse(endpoint)?.newBuilder()?.apply {
+            addQueryParameter("user_id", userId)
+        }?.build().orElse { Errors.network("Incorrect endpoint: $endpoint") }
 
-        val request = Request.Builder()
-            .url(urlBuilder.build())
-            .post(requestBody)
-            .build()
+    //TODO: Maybe add auth data and params to query params instead of body?
```

Comment:
OAuth2 spec sends it as part of the body, so think we'll be leaving it as that for now